snap/quota/quota.go

```diff
@@ -264,18 +268,64 @@ func maxq(a, b quantity.Size) quantity.Size {
 	return b
 }
 
-func (grp *Group) getAllowedCPUs() []int {
+// GetLocalCPUSetQuota returns the current CPU set quota for the group. This
+// does not return any inheritted CPU set quota.
+func (grp *Group) GetLocalCPUSetQuota() []int {
 	if grp.CPULimit == nil || len(grp.CPULimit.AllowedCPUs) == 0 {
 		return []int{}
 	}
 	return grp.CPULimit.AllowedCPUs
 }
 
-func (grp *Group) getTotalCPUPercentage() int {
+// GetCPUSetQuota returns the currently active CPU set quota for this group, which
+// includes the case where the CPU set is inherited from a parent group.
+func (grp *Group) GetCPUSetQuota() []int {
+	localCPUSet := grp.GetLocalCPUSetQuota()
+	if len(localCPUSet) != 0 {
+		return localCPUSet
+	}
+
+	parent := grp.parentGroup
+	for parent != nil {
+		if parent.CPULimit != nil && len(parent.CPULimit.AllowedCPUs) != 0 {
+			return parent.CPULimit.AllowedCPUs
+		}
+		parent = parent.parentGroup
+	}
+	return nil
+}
+
+// GetLocalCPUQuota returns the final calculated count and percentage of the
+// current CPU quota for the group. This does not return any inherited CPU quota, but
+// it does take any inherited CPU set into account to adjust in the case of a relative
+// usage percentage.
+// The following output is expected:
+// Count of 0, 50% - 50% times the number of all allowed cores. This is either
+//                   the full amount of cores present on the system, or it is
+//                   the number of cores allowed for this group.
+// Count of 4, 25% - If the count is provided, it will always be the used multipler
+//                   for the final percentage calculation. This case will be 4x25%.
```

Comment:
Sorry for nitpicking ;) but I find this example confusing in this spot; first, these numbers are not directly related to the logic of this function, are they? second - 0x50% is a special case you mentioned in the PR description, but 4x25% is not, right; or is it special because 4x25=100? I wonder if there a better place for this comment, or if it at the very least the 2nd example should be made less special?

Reply:
Good point, I'll clear this up a bit. As you said, the special case is the first one, not the second one.